File: packages/protocol/src/quorum.ts
// M0 fix (risk #4): 每个阶段需要一个最低响应法定人数（quorum），
// 未达到时该阶段应标记为 partial，而不是让整个 run 失败。

export function computeQuorum(modelCount: number, ratio = 2 / 3): number {
  if (modelCount <= 0) {
    throw new Error("modelCount must be > 0");
  }
  return Math.max(1, Math.ceil(modelCount * ratio));
}

export function meetsQuorum(
  respondentCount: number,
  modelCount: number,
  ratio = 2 / 3
): boolean {
  return respondentCount >= computeQuorum(modelCount, ratio);
}

export interface QuorumCheck {
  met: boolean;
  required: number;
  respondentCount: number;
  partial: boolean;
}

export function checkQuorum(
  respondentCount: number,
  modelCount: number,
  ratio = 2 / 3
): QuorumCheck {
  const required = computeQuorum(modelCount, ratio);
  return {
    met: respondentCount >= required,
    required,
    respondentCount,
    partial: respondentCount < modelCount,
  };
}

File: python/mmd_litellm/quorum.py
from __future__ import annotations

import math

from pydantic import BaseModel


class QuorumCheck(BaseModel):
    met: bool
    required: int
    respondent_count: int
    partial: bool


def compute_quorum(model_count: int, ratio: float = 2 / 3) -> int:
    if model_count <= 0:
        raise ValueError("model_count must be > 0")
    return max(1, math.ceil(model_count * ratio))


def meets_quorum(
    respondent_count: int, model_count: int, ratio: float = 2 / 3
) -> bool:
    return respondent_count >= compute_quorum(model_count, ratio)


def check_quorum(
    respondent_count: int, model_count: int, ratio: float = 2 / 3
) -> QuorumCheck:
    required = compute_quorum(model_count, ratio)
    return QuorumCheck(
        met=respondent_count >= required,
        required=required,
        respondent_count=respondent_count,
        partial=respondent_count < model_count,
    )
